### ISSUE TITLE
Scancode copyright findings are ignored in ReadmeOSS, SPDX, REST API; user copyright findings are ignored in REST API

### ISSUE THREAD

**Issue by alpianon** [fossology/fossology#2700]
### Description

Copyright findings by Scancode are visualized in the WebUI but it seems that they are not output anywhere: if one exports ReadmeOSS or SPDX reports, or if one uses the `/uploads/{id}/item/{itemId}/copyrights` endpoint, such findings are not used at all.

This inconsistency is particularly problematic, because - at least for us - the main reason to use Scancode is its better copyright scanner, but if the corresponding results cannot be used anywhere, the Scancode integration becomes mostly useless.

Moreover, user copyright findings are exported in ReadmeOSS and SPDX, but not when using the REST API.

#### How to reproduce

Upload whatever source package, run only the Scancode copyright scanner (or deactivate all findings by Fossology's copyright scanner), check results in the WebUI (Scancode findings should be there) and add some fake user copyright findings.

Generate and check ReadmeOSS and SPDX reports of the upload, and make REST API calls to ``/uploads/{id}/item/{itemId}/copyrights` for files/items that do have scancode copyright findings and user copyright findigs, and check the response.

The expected results would be to see all copyright findings (Fossology's, Scancode's, User's), but this is not the case.

### Versions

* 4.4.0
* Debian 11

### Where's the problem in the code

REST API: `Fossology\UI\Api\Controllers\CopyrightController\getFileCX()`:   dataType "scancode_statement" (which corresponds to the "scancode_copyright" table) is absent

Report generation: (I did not investigate it yet)

**Comment by alpianon** [comment]
also user copyright findings are not available via REST API. So the latter is a bug, anyway :)

Whatever label, I'm going to fix it anyway, at least for the REST API bug/feature, because I need this to work for my project. I will open a PR, if you are OK with it

**Comment by alpianon** [comment]
@GMishx  @soham4abc I am looking into the REST API code for copyrights; IIUC it partially uses ajax-copyright-hist plugin, but just for read operations, while it directly uses the dao.copyright plugin for update/insert/delete operations.

Is there a particular reason for not calling ajax-copyright-hist plugin methods for all operations instead? It seems to support all operations for all types of findings (Fossology, Scancode, User decisions)

**Comment by dvjsharma** [comment]
Hello @alpianon, if you closely examine the `ajax-copyright-hist` plugin code, you will see that it calls the `updateTable` method (except for getting the copyright info) for any kind of modification, which is provided by the copyright DAO itself. Hence in the controller script, DAO is used directly. 

As for the API issue, after analysing the code, it appears that we lack the necessary framework to retrieve and update/restore/delete user findings in copyrights. Currently, it solely manages scanner findings. I particularly looked into getting the user findings along with scanner findings, below is the SQL query (general case) which is currently in use to get all copyrights. (source : [ajax-copyright-hist.php](https://github.com/fossology/fossology/blob/e1b862b165b4666a6e3a46d7972fa67d7c774f2a/src/copyright/ui/ajax-copyright-hist.php#L231-L309))
```SQL
SELECT 
    mcontent AS content, 
    mhash AS hash, 
    copyright_count 
FROM 
    (
        SELECT
            (CASE 
                WHEN (ce.content IS NULL OR ce.content = '') THEN cp.content 
                ELSE ce.content 
            END) AS mcontent,
            (CASE 
                WHEN (ce.hash IS NULL OR ce.hash = '') THEN cp.hash 
                ELSE ce.hash 
            END) AS mhash,
            count(*) AS copyright_count 
        FROM 
            copyright AS cp 
        INNER JOIN 
            uploadtree_a AS UT ON cp.pfile_fk = UT.pfile_fk 
        LEFT JOIN 
            copyright_event AS ce ON ce.copyright_fk = cp.copyright_pk 
                                    AND ce.upload_fk = $5 
                                    AND ce.uploadtree_fk = UT.uploadtree_pk 
        WHERE 
            cp.content != '' 
            AND (UT.lft BETWEEN $1 AND $2) 
            AND cp.type = $3 
            AND cp.agent_fk = ANY($4::int[]) 
            AND (ce.is_enabled IS NULL OR ce.is_enabled = 'true') 
            AND UT.upload_fk = $5   
        GROUP BY 
            mcontent, 
            mhash
    ) AS k    
OFFSET 
    $6 
LIMIT 
    $7;
```
This query overlooks user findings stored in the `copyright_decision` table as shown in the image below:
![image](https://github.com/fossology/fossology/assets/119073504/cc3e3fa8-6173-4838-b304-4470ea0db78d)
I updated the query to:
```SQL
SELECT 
    mcontent AS content, 
    mhash AS hash, 
    copyright_count 
FROM 
    (
        SELECT
            (CASE 
                WHEN (ce.content IS NULL OR ce.content = '') THEN cp.content 
                ELSE ce.content 
            END) AS mcontent,
            (CASE 
                WHEN (ce.hash IS NULL OR ce.hash = '') THEN cp.hash 
                ELSE ce.hash 
            END) AS mhash,
            count(*) AS copyright_count 
        FROM 
            copyright AS cp 
        INNER JOIN 
            uploadtree_a AS UT ON cp.pfile_fk = UT.pfile_fk 
        LEFT JOIN 
            copyright_event AS ce ON ce.copyright_fk = cp.copyright_pk 
                                    AND ce.upload_fk = $5
                                    AND ce.uploadtree_fk = UT.uploadtree_pk 
        WHERE 
            cp.content != '' 
            AND (UT.lft BETWEEN $1 AND $2) 
            AND cp.type = $3 
            AND cp.agent_fk = ANY($4::int[]) 
            AND (ce.is_enabled IS NULL OR ce.is_enabled = 'true') 
            AND UT.upload_fk = $5  
        GROUP BY 
            mcontent, 
            mhash

        UNION ALL
       -- Considered results form copyright_decision table as well
        SELECT
            cd.textfinding AS mcontent,
            cpd.hash AS mhash,
            COUNT(DISTINCT cd.copyright_decision_pk) AS copyright_count
        FROM 
            copyright_decision AS cd 
        INNER JOIN 
            pfile AS PF ON cd.pfile_fk = PF.pfile_pk 
        INNER JOIN 
            uploadtree_a AS UT ON PF.pfile_pk = UT.pfile_fk 
        LEFT JOIN 
            copyright AS cp ON cp.pfile_fk = PF.pfile_pk
                               AND (cp.is_enabled IS NULL OR cp.is_enabled = 'true')
                               AND UT.upload_fk = $5
        LEFT JOIN
            copyright_decision AS cpd ON cd.textfinding = cpd.textfinding
                                         AND UT.upload_fk = $5
        WHERE 
            UT.lft BETWEEN $1 AND $2
            AND (cd.is_enabled IS NULL OR cd.is_enabled = 'true') 
            AND UT.upload_fk = $5
        GROUP BY 
            cd.textfinding, cpd.hash
    ) AS k    
OFFSET 
    $6
LIMIT 
    $7;
```
And I was able to obtain the expected combined results from the scanner and user findings:
![image](https://github.com/fossology/fossology/assets/119073504/7c10e8cc-c1c5-49dd-9a75-4a0082c06a15)

This approach can be used to modify all the `GET` operations to include user findings namely,`getFileCopyrights` and `getTotalFileCopyrights` (need to return the count which includes user findings as well- _already made the [query](https://pastebin.com/GqXQpAGU)_).

![image](https://github.com/fossology/fossology/assets/119073504/03528b1f-21fb-447b-95c5-e5d95d51a25d)
_result for `getTotalFileCopyrights`_

For all update/restore/delete operations, we use a shared method `updateTable` provided by [Copyright DAO](https://github.com/fossology/fossology/blob/e1b862b165b4666a6e3a46d7972fa67d7c774f2a/src/lib/php/Dao/CopyrightDao.php#L448-L534).It can be modified similarly to access user findings from the `copygirht_decision` table for modifications. It also saves updated outcomes in the `copyright_event` table, which I am currently researching on how to manage that situation. I'll create a PR once I finish the modifications. Please confirm if the approach is acceptable.

CC @GMishx @shaheemazmalmmd

**Comment by GMishx** [comment]
Hey @dvjsharma , IMO, the data from scanner findings, user findings, and findings from external agents (like scancode) should not be mixed (at least in the UI). Thus, it will be better to create separate endpoints for all 3 kinds. WDYT?

**Comment by dvjsharma** [comment]
Agreed @GMishx, creating separate endpoints for the 3 types of findings would improve clarity and integration. I will look into this. Until then, maybe the `/uploads/{id}/copyrights` endpoint could be used as it also presents user findings.

CC @alpianon

**Comment by alpianon** [comment]
Having separate endpoints, keeping the current ones only for internal scanner findings,  would make sense; alternatively, an additional parameter may be added to choose between internal scanner, scancode and user findings, defaulting to internal scanner.

In both cases, backwards compatibility would be ensured for those who may already have built some integration using the existing endpoints

**Comment by KEO-RSP** [comment]
I also ran into this in release 4.6.0 and would be happy if this becomes more streamlined / integrated in the future. If not, it might make sense to document the current behaviour or show some sort of hint in the tabbed copyright findings view.
